### PR TITLE
Lock player move

### PR DESCRIPTION
**Pull request by john0312**

### CODE REVIEW

File: services/gateway/gateway-service.mjs [hunk 1]
```diff
@@ -368,23 +377,27 @@ class GatewayService {
...
-    const ret = await this._enterCoord(msg.mapCoord);
-
-    // If the player was in ghost mode, ignore the occupation check.
-    if (!(socket.playerData.ghostMode || allowOverlap)) {
-      // If the player moves in normal mode, check the occupation.
-      if (!msg.ghostMode && !ret) {
-        await this._leaveCoord(msg.mapCoord);
-        return false;
+    let res;
+    res = await socket.moveLock.acquire('move', async () => {
```
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:381]: `const res = await ...`

File: services/gateway/gateway-service.mjs [hunk 2]
```diff
@@ -210,77 +211,82 @@ class GatewayService {
       }
     }
 
-    // Load the player data.
-    socket.playerData = await this.dir.getPlayerData(playerID);
-    socket.playerID = playerID;
+    // Acquire this lock if you need to move the player.
+    socket.moveLock = new AsyncLock();
+    await socket.moveLock.acquire('move', async () => {
```
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:216]: I think it is overkill to make the whole process locked.
The necessary part, i.e. the code that really deals with moving players, is only `socket.on('avatarSelect', ...`
Note that I suggest a little refactoring in below comment.

File: services/gateway/gateway-service.mjs [hunk 3]
```diff
@@ -210,77 +211,82 @@ class GatewayService {
...
-      this.onPlayerUpdate(socket, PlayerSyncMessage.fromObject(msg));
-      // onPlayerUpdate is async, so returns immediately.
-    });
+      // Synchronize the state.
+      const initLoc = socket.playerData.mapCoord ?? this.gameMap.getRandomSpawnPoint();
+      await this._enterCoord(initLoc);
+      socket.playerData.mapCoord = initLoc;
+      socket.playerData.facing = 'D';
+      socket.playerData.lastMovingTime = Date.now();
+      socket.playerData.ghostMode = false;
```
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:262]: Maybe move these lines into `socket.on('avatarSelect' ...`?

**Review by ktpss95112 — COMMENTED** [review]
In function `addSocket`, the critical section is only from `this._enterCoord()` to `this._broadcastPlayerUpdate()`. Maybe consider only lock these code instead of almost the whole function.

File: services/gateway/gateway-service.mjs (hunk 1 above)
Comment (john0312) [reply services/gateway/gateway-service.mjs]: Done.

File: services/gateway/gateway-service.mjs (hunk 2 above)
Comment (john0312) [reply services/gateway/gateway-service.mjs]: Added comments on why that is done. (To serialize the connect and disconnect handlers.)

File: services/gateway/gateway-service.mjs (hunk 3 above)
Comment (john0312) [reply services/gateway/gateway-service.mjs]: Done.

**Comment by john0312** [comment]
Reviews have been addressed. There are still some race conditions in this code (should've locked handler for avatarSelect). However, those will be fixed in the coming PR.

**Review by ktpss95112 — APPROVED** [review]

**Pull request merged**